Fix responsive styles and tweak prev/next links

renatho opened the pull request:
Fixes responsive design issues.
Fixes #4449

### Changes proposed in this Pull Request

* It fixes the responsive design issues in the learning mode.
* It updates the prev/next button to be displayed as disabled when it's in the first/last lesson.
* It moves the prev/next classes to a single class.
* It improves accessibility by hiding the footer too when collapsing sidebar.
* It updates the "Expand" button to be full width with a background to avoid being over the content.

### Testing instructions

<!--
If the functionality of this PR is behind a feature flag, please include the
following testing step, using the correct name for the feature flag. (If you
aren't sure, just ignore this step)

* Enable the feature flag: `add_filter( 'sensei_feature_flag_{THE_FLAG_NAME}', '__return_true' );`
-->
* Create a course with many lessons.
* Enable the learning mode for this course.
* Go to the learning mode as a student and as an admin, resize the window to different sizes, open and close the menu, and make sure everything works properly.
* Also make sure you can navigate properly through the Prev/Next buttons on the top bar.
* Navigate to the first lesson, and make sure you see the Previous link as disabled.
* Navigate to the last lesson, and make sure you see the Next link as disabled.

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video

Video showing the responsive design (The part with the inspector open was only to see it working wider than `600px` and smaller than `600px`):

https://user-images.githubusercontent.com/876340/151233011-3ba489f1-9935-449b-b93e-cc52bdb3b505.mp4

<img width="354" alt="Screen Shot 2022-01-25 at 16 47 09" src="https://user-images.githubusercontent.com/876340/151048730-5e7a6801-4334-4165-bca2-d1dd6e06dcdc.png">

renatho commented on includes/blocks/course-theme/class-prev-lesson.php:
```diff
@@ -47,23 +47,29 @@ public function render( array $attributes = [] ) : string {
 		}
 
-		$urls = sensei_get_prev_next_lessons( $lesson_id );
+		$urls           = sensei_get_prev_next_lessons( $lesson_id );
```
Prev/next buttons contain some duplication code. Since we plan to [refactor that](https://github.com/Automattic/sensei/issues/4449), I kept it for now.

yscik commented on assets/course-theme/adminbar-layout.js:
```diff
@@ -15,6 +20,15 @@ const trackAdminbarOffset = () => {
+	 * 2. The admin bar contains an animated transition, so this transition
+	 *    needs to be completed in order to make the correct calc.
+	 */
+	// eslint-disable-next-line @wordpress/no-global-event-listener
+	window.addEventListener( 'resize', debounce( updateAdminbarOffset, 500 ) );
```
I'm not sure we should care about resizing the window. We often test responsiveness that way, but the real-world use case is just people having a smaller screen. AFAIK `lodash` doesn't support tree-shaking so for this debounce we're bringing that entire library to our frontend, which is probably not worth it.

dadish replied:
I agree. I don't think it's worth supporting a resize change. I don't think anyone resizes their browser after visiting a website. Only developers who need to support responsive layout do that. 😆

renatho replied:
Good catch! It wasn't in our bundle, but it was being [included through WP lodash](https://www.npmjs.com/package/@wordpress/dependency-extraction-webpack-plugin). But yeah! We don't need to import it there. Updated it here: https://github.com/Automattic/sensei/pull/4684/commits/836dae5c4422b49c8d37e0dda023f1672470f298

![Screen Shot 2022-01-27 at 11 04 41](https://user-images.githubusercontent.com/876340/151374201-ed1a3e45-0cd0-46e0-9464-b8831a933f91.png)

I just don't agree that we shouldn't fix it for resize, since it's just one extra event not introducing a big complexity. I think if we can fix it easily, we should do it, and with that we avoid weird issues if the user decides to resize the window. Unless someone see any con on that, I think we should keep that.

> If we end up keeping a handler on resize, adding the { capture: false, passive: true } options to the addEventListener might improve performance, making it kind of like a non-blocking call. (Not sure if it's relevant for the resize event)

Since that documentation is specifically about the scrolling, I'll not add that for now. Probably it won't be a big deal since we're doing that with the debounce on the resize. But if we notice any performance issue, we can try that. ;)

dadish replied:
> with that we avoid weird issues if the user decides to resize the window.

On the other hand, if we make this page resilient to window resize, then user will expect all pages to behave this way. So they will go ahead and complain if other pages break when the window is resized. 🤷 

I think it's better to fix it if/when we have real complains about it. Because otherwise we are adding logic based on our own speculation instead of real world usage data.

renatho replied:
> On the other hand, if we make this page resilient to window resize, then user will expect all pages to behave this way. So they will go ahead and complain if other pages break when the window is resized. 🤷

I don't think it's a problem because other pages don't have any logic basic on that. It's something specific we have in the Learning Mode.

> I think it's better to fix it if/when we have real complains about it. Because otherwise we are adding logic based on our own speculation instead of real world usage data.

If it would introduce some big complexity to the code, I could agree with you, but considering it causes a real issue, and it's very easy to fix (just registering an extra event), I don't see a reason to not fix it. 🤔

dadish replied:
Yeah, I think you're right. It wouldn't hurt us to add this. 👍

yscik commented on assets/css/sensei-course-theme/focus-mode.scss:
```diff
@@ -1,92 +1,113 @@
+			&__focus-mode-toggle {
+				top: 0;
+				left: 400px;
+				width: 100vw;
```
Instead of the full-width expand button with a background, what if we ditch the label, and just show the icon
<img width="17" alt="image" src="https://user-images.githubusercontent.com/176949/151460910-4c200cc4-a0e2-4050-ae59-8e45c6fcc108.png"> in the top right corner when collapsed? There is padding enough around the content that it won't overlap, and the user already interacted it with it so won't need an extra explanation. (Just add 'Expand' as a `title` attribute)

renatho replied:
With this approach, the button seems a little weird! But I liked your other idea about a button in the header. See the other comment https://github.com/Automattic/sensei/pull/4684#discussion_r794092400.

<img width="782" alt="Screen Shot 2022-01-28 at 10 31 07" src="https://user-images.githubusercontent.com/876340/151560740-e45fcc7c-9614-48e6-9de8-64125da5f22e.png">

yscik commented on assets/css/sensei-course-theme/layout.scss:
```diff
@@ -1,13 +1,17 @@
 :root {
-	--header-height: var(--focus-header-height, 75px);
+	--header-height: 75px;
```
Title wrapping is broken for me on this branch, it breaks into a new line even with ellipsis working:

<img width="397" alt="image" src="https://user-images.githubusercontent.com/176949/151461556-103f9b21-9e3f-4432-84e0-92e82222039f.png">

_(Not related to the line I commented this on, just wanted to start a thread)_

renatho replied:
Interesting! For mobile, it should hide the progress based on [our styles](https://github.com/Automattic/sensei/blob/74225c7e3ce1c3877c56366552b54dffa9c27646/assets/css/sensei-course-theme/mobile.scss#L55-L57). I couldn't reproduce it on my end. Could you inspect on your end for me to understand what happened there? Maybe the block was moved to some part where the class didn't get?

yscik replied:
This is when the sidebar is open. But the problem is even when it's closed, the title disappears because it breaks into a new line. I'll look into it.

renatho replied:
Ah, okay! I got it! I can see it now! =)

renatho replied:
It should be fixed now with this commit: https://github.com/Automattic/sensei/pull/4684/commits/0f4bfa73365659c5c1481e0b2ab684878d346d45

Do you remember why there was that property there? Maybe we just forgot to remove it before some change?

yscik replied:
I think it's for wrapping the course progress text block to a new line: we show that when the sidebar is open:

<img width="547" alt="image" src="https://user-images.githubusercontent.com/176949/151566901-61ade82e-57d4-468b-a1c2-91f48d2a464b.png">

yscik commented on assets/css/sensei-course-theme/mobile.scss:
```diff
@@ -23,6 +17,7 @@
 
 		&--sidebar-open {
 			--header-height: 90px;
+			overflow: hidden;
```
Just noting that we can move the Focus mode toggle block out of the sidebar if it makes life easier, it's only in there to look a bit more sensible in the editor.
It could even make sense as an up/down arrow in the header, left to the logo:

<img width="1208" alt="image" src="https://user-images.githubusercontent.com/176949/151462180-2a146492-864e-4d41-a8ae-181a0c65ee8e.png">

renatho replied:
I liked this idea! Implemented it here: https://github.com/Automattic/sensei/pull/4684/commits/ec632005c8f0779121be28414d7ebfac66fbf166

https://user-images.githubusercontent.com/876340/151560799-d80aaaca-6b82-4522-b86c-98d23664a605.mov

renatho commented:
Since it's a relevant change from [here](https://github.com/Automattic/sensei/pull/4684#discussion_r794092400), I'm adding it as a comment for documentation of this behavior:

https://user-images.githubusercontent.com/876340/151560799-d80aaaca-6b82-4522-b86c-98d23664a605.mov

renatho commented on assets/css/sensei-course-theme/focus-mode.scss:
```diff
@@ -1,92 +1,76 @@
-				display: none;
-			}
-			.wp-block-site-logo {
-				margin-right: -12px;
```
@yscik , let me know if I missed anything here. I changed it to `display: none` directly to avoid the `gap` and avoid fixing it with the negative margin.

yscik replied:
I think it was done that way for animation reasons, looks good with your solution!
One bit, for the title, I think we should remove the transition of the font-size, since it's lagging behind. Or just need to adjust the timing/easing?

renatho replied:
Good catch! It was transitioning the `h2` and the `a`. And weird that it was increasing the duration. But removed from the `a` and it's fixed now. https://github.com/Automattic/sensei/pull/4684/commits/42ccb7c1e5b3635d92c0a4f182411dd51a6232a9

yscik reviewed (approved):
Looks great! I added a quick commit to remove the Expand/Collapse labels we no longer use: 217abe4949a49d8d8792421b89e90df440979c70

And also needed to order the media queries a bit in 2e5df090a571dcb5c3e455f32afaa5fa0dc2e039 to fix this transition:
<img src="https://user-images.githubusercontent.com/176949/151587976-b5fb5b65-def9-44d6-998b-44cd901c1096.gif" width=400 />

renatho commented:
> I added a quick commit to remove the Expand/Collapse labels we no longer use: 217abe4

It was there for accessibility purposes. But with the title probably not super necessary. I added another commit just removing some other unused classes. https://github.com/Automattic/sensei/pull/4684/commits/3fb32a22b4464678a04f85d8e4047fc0b6540a86

Since my last commit was only removing unused classes, I'm merging it.

yscik commented:
Yeah I think the `Toggle` label is enough to understand. Good work with it all!
